Wrappers/Python/cil/optimisation/operators/Operator.py

```diff
@@ -216,107 +250,105 @@ def PowerMethod(operator, max_iteration=10, initial=None, tolerance = 1e-5,  ret
         diff = numpy.finfo('d').max
         i = 0
         while (i < max_iteration and diff > tolerance):
-            i+=1
-            
-            operator.direct(x0, out = y_tmp)
+            operator.direct(x0, out=y_tmp)
 
-            if symmetric:                
-                #swap datacontainer references
+            if method=="direct_only":
+                # swap datacontainer references
                 tmp = x0
                 x0 = y_tmp
                 y_tmp = tmp
-            else:
-                operator.adjoint(y_tmp,out=x0)
-            
+            elif method=="composed_with_adjoint":
+                operator.adjoint(y_tmp, out=x0)
+
             # Get eigenvalue using Rayleigh quotient: denominator=1, due to normalization
             x0_norm = x0.norm()
+            if x0_norm < tolerance:
+                Warning(
+                    'The operator has at least one zero eigenvector and is likely to be nilpotent')
+                eig_new = 0.
+                break
             x0 /= x0_norm
 
-            eig_new =  numpy.abs(x0_norm)
-            if not symmetric:
+            eig_new = numpy.abs(x0_norm)
+            if method=="composed_with_adjoint":
                 eig_new = numpy.sqrt(eig_new)
-
             diff = numpy.abs(eig_new - eig_old)
-            eig_list.append(eig_new)   
-            eig_old = eig_new       
+            eig_list.append(eig_new)

```

Comment:
This should only be created and appended too when it's been requested by `return_all`. I don't think it is used otherwise.